File: spring-boot-ecommerce/src/main/java/com/springboot/ecommerce/service/CheckoutService.java
package com.springboot.ecommerce.service;

import com.springboot.ecommerce.dto.PaymentInfo;
import com.springboot.ecommerce.dto.Purchase;
import com.springboot.ecommerce.dto.PurchaseResponse;
import com.stripe.exception.StripeException;
import com.stripe.model.PaymentIntent;

public interface CheckoutService {
    PurchaseResponse placeOrder(Purchase purchase);
    PaymentIntent createPaymentIntent(PaymentInfo paymentInfo) throws StripeException;
}

File: angular-ecommerce/src/app/services/checkout.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { BehaviorSubject, Observable, Subject } from 'rxjs';
import { environment } from 'src/environments/environment';
import { Purchase } from '../models/purchase';
import { PaymentInfo } from '../models/payment-info.model';

@Injectable({
  providedIn: 'root'
})
export class CheckoutService {
  private purchaseUrl = environment.ecommerceApiUrl + '/checkout/purchase';
  private paymentIntentUrl = environment.ecommerceApiUrl + '/checkout/payment-intent';

  orderTrackingNumberNotification: Subject<string> = new BehaviorSubject('');

  constructor(private http: HttpClient) {}

  placeOrder(purchase: Purchase): Observable<any> {
    return this.http.post<Purchase>(this.purchaseUrl, purchase);
  }

  createPaymentIntent(paymentInfo: PaymentInfo): Observable<any> {
    return this.http.post<PaymentInfo>(this.paymentIntentUrl, paymentInfo);
  }
}
